subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,673 @@
-            <para>
-                The task's <literal>inputs</literal> property is of type <apilink class="org.gradle.api.tasks.TaskInputs"/>.
-                The task's <literal>outputs</literal> property is of type <apilink class="org.gradle.api.tasks.TaskOutputs"/>.
+            <para>As part of incremental build, Gradle tests whether any of the task inputs or outputs
+                have changed since the last build. If they haven’t, Gradle can consider the task up to
+                date and therefore skip it. Also note that incremental build won’t work unless a task
+                has at least one task output, although tasks usually have at least one input as well.
             </para>
-            <para>
-                A task with no defined outputs will <emphasis>never</emphasis> be considered up-to-date.
-                For scenarios where the outputs of a task are not files, or for more complex scenarios, the
-                <apilink class="org.gradle.api.tasks.TaskOutputs" method="upToDateWhen(groovy.lang.Closure)" /> method allows you to calculate programmatically if
-                the tasks outputs should be considered up to date.
+            <para>What this means for build authors is simple: you need to tell Gradle which task
+                properties are inputs and which are outputs. If a task property affects the output, be
+                sure to register it as an input, otherwise the task will be considered up to date when
+                it’s not. Conversely, don’t register properties as inputs if they don’t affect the
+                output, otherwise the task will potentially execute when it doesn’t need to. Also be
+                careful of non-deterministic tasks that may generate different output for exactly the
+                same inputs: these should not be configured for incremental build as the up-to-date
+                checks won’t work.
             </para>
-            <para>
-                A task with only outputs defined will be considered up-to-date if those outputs are unchanged since the previous build.
+            <para>Let’s now look at how you can register task properties as inputs and outputs.</para>
+
+            <section id="sec:task_input_output_annotations">
+                <title>Custom task types</title>
+                <para>If you’re implementing a custom task as a class, then it takes just two steps to
+                    make it work with incremental build:
+                </para>
+                <orderedlist>
+                    <listitem>Create typed fields or properties (via getter methods) for each of your
+                        task inputs and outputs</listitem>
+                    <listitem>Add the appropriate annotation to each of those fields and getter methods
+                    </listitem>
+                </orderedlist>
+                <para>Gradle supports three main categories of input and output:</para>
```

Comment:
I think both may be acceptable (it's hard to find an authoritative resource on this), but I'll change it.